### ISSUE TITLE
Prepare for PyTorch 0.4.0

### ISSUE THREAD

**Issue by wohlert** [wohlert/semi-supervised-pytorch#5]
## Loss functions

Many loss functions in the code have manual implementations such as `binary_cross_entropy` and `cross_entropy`, rather than their PyTorch counterparts. This is due to PyTorch not supporting [losses per batchelement](https://github.com/pytorch/pytorch/issues/264) as of 0.3.0. For 0.4.0, the following should be possible.

Current implementation:
```python
def binary_cross_entropy(r, x):
    return -torch.sum(x * torch.log(r + 1e-8) + (1 - x) * torch.log(1 - r + 1e-8), dim=-1)
```

Future:
```python
torch.nn.functional.binary_cross_entropy(r, x, size_average=False, reduce=False)
```

## Half-precision

Many models benefit from using half-precision accuracy on the GPU. At the moment, the PyTorch API is cumbersome with fp16 and it is causing problems with batchnorm. However, it could be a possibility in the future (see https://github.com/csarofeen/examples/tree/fp16_examples_cuDNN-ATen/imagenet).

Future:
```python
if cuda:
    torch.set_default_tensor_type('torch.cuda.HalfTensor')
model.cuda().half()
```
